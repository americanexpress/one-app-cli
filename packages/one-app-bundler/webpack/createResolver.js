/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const {
  NodeJsInputFileSystem,
  CachedInputFileSystem,
  ResolverFactory,
} = require('enhanced-resolve');

// This enhanced resolver ensures that the code for a lib that is exposed is the
// same code that is used within the bundle. This bug would happen if an exposed
// lib had a module or browser field in addition to the main field in its
// package.json.

module.exports = function createResolver({ mainFields, resolveToContext = false }) {
  const enhancedResolver = ResolverFactory.createResolver({
    fileSystem: new CachedInputFileSystem(new NodeJsInputFileSystem(), 4000),
    useSyncFileSystemCalls: true,
    extensions: ['.js', '.jsx', '.json'],
    mainFields,
    resolveToContext,
  });

  const trailingSlash = process.platform === 'win32' ? '\\' : '/';
  // if resolveToContext add a trailing slash to indicate the value is a folder rather than  a file
  return (request) => `${enhancedResolver.resolveSync({}, __dirname, request)}${resolveToContext ? trailingSlash : ''}`;
};
