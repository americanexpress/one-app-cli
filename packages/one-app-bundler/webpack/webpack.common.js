/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/**
 * Webpack Base Configuration
 */

const webpack = require('webpack');
const TerserPlugin = require('terser-webpack-plugin');
const ESLintPlugin = require('eslint-webpack-plugin');
const validateNodeEnv = require('../utils/validateNodeEnv');

validateNodeEnv();

const nodeEnvIsProduction = process.env.NODE_ENV === 'production';

const prodPlugins = [
  new webpack.LoaderOptionsPlugin({
    minimize: true,
    debug: false,
  }),
];

const plugins = [
  new webpack.EnvironmentPlugin([
    'NODE_ENV',
  ]),
  new ESLintPlugin({
    allowInlineConfig: false,
    ignore: false,
    failOnError: true,
    useEslintrc: false,
    overrideConfig: {
      parserOptions: {
        ecmaVersion: 6,
        sourceType: 'module',
        ecmaFeatures: { modules: true },
      },
      rules: {
        'no-restricted-modules': ['error', 'create-react-class'],
        'no-restricted-imports': ['error', 'create-react-class'],
        '@americanexpress/one-app/no-app-config-on-client': 'error',
      },
      plugins: ['@americanexpress/one-app'],
    },
  }),
  ...(nodeEnvIsProduction ? prodPlugins : []),
];

module.exports = {
  profile: true,
  devtool: nodeEnvIsProduction ? false : 'source-map',
  optimization: {
    minimize: nodeEnvIsProduction,
    minimizer: [
      new TerserPlugin({
        test: /\.jsx?$/i,
        extractComments: false,
        terserOptions: {
          compress: {
            drop_console: true,
          },
          keep_fnames: true,
        },
      }),
    ],
  },
  mode: nodeEnvIsProduction ? 'production' : 'development',
  module: {
    rules: [
      {
        test: /\.(ttf|eot|svg|png|jpg)(\?.*)?$/,
        use: [{ loader: 'url-loader' }],
      },
      {
        test: /\.(woff|woff2)(\?.*)?$/,
        use: [{
          loader: 'url-loader',
          options: {
            limit: 10000,
            mimetype: 'application/font-woff',
          },
        }],
      },
    ],
  },
  externals: {
    '@americanexpress/one-app-router': {
      var: 'OneAppRouter',
      commonjs2: '@americanexpress/one-app-router',
    },
    'create-shared-react-context': {
      var: 'CreateSharedReactContext',
      commonjs2: 'create-shared-react-context',
    },
    holocron: {
      var: 'Holocron',
      commonjs2: 'holocron',
    },
    react: {
      var: 'React',
      commonjs2: 'react',
    },
    'react-dom': {
      var: 'ReactDOM',
      commonjs2: 'react-dom',
    },
    redux: {
      var: 'Redux',
      commonjs2: 'redux',
    },
    'react-redux': {
      var: 'ReactRedux',
      commonjs2: 'react-redux',
    },
    reselect: {
      var: 'Reselect',
      commonjs2: 'reselect',
    },
    immutable: {
      var: 'Immutable',
      commonjs2: 'immutable',
    },
    '@americanexpress/one-app-ducks': {
      var: 'OneAppDucks',
      commonjs2: '@americanexpress/one-app-ducks',
    },
    'holocron-module-route': {
      var: 'HolocronModuleRoute',
      commonjs2: 'holocron-module-route',
    },
    'prop-types': {
      var: 'PropTypes',
      commonjs2: 'prop-types',
    },
    'react-helmet': {
      var: 'ReactHelmet',
      commonjs2: 'react-helmet',
    },
  },
  plugins,
};
