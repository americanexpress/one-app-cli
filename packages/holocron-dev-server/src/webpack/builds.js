/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import webpack from 'webpack';

import { createHolocronModuleWebpackConfig, createExternalsDllWebpackConfig } from './configs/bundles';
import { modulesLibraryVarName } from './helpers';
import { ufs } from '../utils';
import {
  logError,
  logWebpackStatsWhenDone,
  logWhenWebpackInvalid,
  logExternalsBuilding,
} from '../utils/logs';

export function buildWebpackConfig(config) {
  return new Promise((resolve, reject) => {
    webpack(config).run((err, stats) => {
      if (err) {
        logError(err);
        reject(err);
      } else {
        resolve(stats);
      }
    });
  });
}

export function buildModuleExternalsDllBundle(config = {}) {
  const { externals: externalsToPackageDll = [] } = config;

  if (!(externalsToPackageDll.length > 0)) return Promise.resolve();

  return buildWebpackConfig(
    createExternalsDllWebpackConfig({
      entries: externalsToPackageDll,
    })
  ).then((stats) => {
    logExternalsBuilding(externalsToPackageDll);
    return stats;
  });
}

export function createHotHolocronCompiler({
  modules,
  externals,
  environmentVariables,
  webpackConfigPath,
}) {
  const holocronWebpackConfig = createHolocronModuleWebpackConfig({
    hot: true,
    modules,
    externals,
    environmentVariables,
    webpackConfigPath,
  });
  const compiler = webpack(holocronWebpackConfig);
  compiler.inputFileSystem = ufs;
  compiler.resolvers.normal.fileSystem = compiler.inputFileSystem;
  compiler.resolvers.context.fileSystem = compiler.inputFileSystem;
  compiler.hooks.done.tap(modulesLibraryVarName, logWebpackStatsWhenDone);
  compiler.hooks.invalid.tap(modulesLibraryVarName, logWhenWebpackInvalid);
  return compiler;
}
