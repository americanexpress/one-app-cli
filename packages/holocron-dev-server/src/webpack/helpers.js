/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { version } from 'webpack';

export const modulesLibraryVarName = '__holocron_modules__';
export const externalsLibraryVarName = '__externals__';

export const modulesFilename = '[name]/[name].js';
export const assetModuleFilename = 'assets/[name].[ext]';

export const jsxTest = /\.jsx?$/i;
export const fileTest = /\.(woff|woff2|ttf|eot|svg|png|jpg|jpeg|gif|webm)(\?.*)?$/;
export const cssTest = /\.(sa|sc|c)ss$/;
export const nodeModulesPattern = /node_modules/;

export function getWebpackVersion() {
  return parseInt(version, 10);
}

export function createExternalEntry([packageName, varName]) {
  return {
    [packageName]: {
      commonjs2: packageName,
      ...varName
        ? {
          var: varName,
          root: varName,
        }
        : {},
    },
  };
}

export function createOneAppExternals(additionalExternals = []) {
  // TODO: pull this from one-app-bundler when merging
  return [
    ['@americanexpress/one-app-ducks', 'OneAppDucks'],
    ['@americanexpress/one-app-router', 'OneAppRouter'],
    ['create-shared-react-context', 'CreateSharedReactContext'],
    ['holocron', 'Holocron'],
    ['holocron-module-route', 'HolocronModuleRoute'],
    ['immutable', 'Immutable'],
    ['prop-types', 'PropTypes'],
    ['react', 'React'],
    ['react-dom', 'ReactDOM'],
    ['react-helmet', 'ReactHelmet'],
    ['react-redux', 'ReactRedux'],
    ['redux', 'Redux'],
    ['reselect', 'Reselect'],
    ...additionalExternals.map((external) => (Array.isArray(external) ? external : [external])),
  ]
    .map(createExternalEntry)
    .reduce((map, next) => ({ ...map, ...next }), {});
}

const createHotModuleEntry = ({ moduleName, modulePath }, hot) => ({
  [moduleName]: [
    ...hot
      ? [require.resolve('webpack-hot-middleware/client'), require.resolve('react-refresh/runtime')]
      : [],
    `${modulePath}/src/index.js`,
  ],
});

export function createHolocronModuleEntries({ modules = [], hot = false } = {}) {
  return modules
    .map((module) => createHotModuleEntry(module, hot))
    .reduce((map, next) => ({ ...map, ...next }), {});
}
