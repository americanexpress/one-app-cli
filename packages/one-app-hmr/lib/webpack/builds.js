/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import webpack from 'webpack';

import { createHolocronModuleWebpackConfig, createExternalsDllWebpackConfig } from './configs';
import { modulesLibraryVarName } from './helpers';
import { ufs } from '../utils';
import {
  error,
  logWebpackStatsWhenDone,
  logWhenWebpackInvalid,
  logExternalsBuilding,
} from '../utils/logs';

export function runWebpackCompiler(compiler) {
  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) {
        error(err);
        reject(err);
      } else {
        resolve(stats);
      }
    });
  });
}

export function buildWebpackConfig(config) {
  return runWebpackCompiler(webpack(config));
}

export function buildModuleExternalsDllBundle(config = {}) {
  const { externals: externalsToPackageDll = [], name } = config;

  if (!(externalsToPackageDll.length > 0)) return Promise.resolve();

  return buildWebpackConfig(
    createExternalsDllWebpackConfig({
      name,
      entries: externalsToPackageDll,
    })
  ).then((stats) => {
    logExternalsBuilding(externalsToPackageDll);
    return stats;
  });
}

export function createHotHolocronCompiler({
  modules,
  externals,
  environmentVariables,
  babelConfig,
  webpackConfigPath,
}) {
  const holocronWebpackConfig = createHolocronModuleWebpackConfig({
    hot: true,
    modules,
    externals,
    environmentVariables,
    babelConfig,
    webpackConfigPath,
  });
  const compiler = webpack(holocronWebpackConfig);
  compiler.inputFileSystem = ufs;
  compiler.resolvers.normal.fileSystem = compiler.inputFileSystem;
  compiler.resolvers.context.fileSystem = compiler.inputFileSystem;
  compiler.hooks.done.tap(modulesLibraryVarName, logWebpackStatsWhenDone);
  compiler.hooks.invalid.tap(modulesLibraryVarName, logWhenWebpackInvalid);
  return compiler;
}
