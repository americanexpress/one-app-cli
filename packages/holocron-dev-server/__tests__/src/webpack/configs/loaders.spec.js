/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import {
  fileLoader,
  cssLoader,
  jsxLoader,
  createJavaScriptSourceLoadersConfigFragment,
  createEsBuildConfigFragment,
} from '../../../../src/webpack/configs/loaders';
import { getWebpackVersion } from '../../../../src/webpack/helpers';

jest.mock('../../../../src/webpack/helpers', () => {
  const originalModule = jest.requireActual('../../../../src/webpack/helpers');

  return {
    ...originalModule,
    getWebpackVersion: jest.fn(() => 4),
  };
});

describe('fileLoader', () => {
  test('returns loader config for file types for webpack version 4 and below', () => {
    expect(fileLoader()).toMatchSnapshot();
  });
  test('returns loader config for file types for webpack version 5 and above', () => {
    getWebpackVersion.mockImplementationOnce(() => 5);

    expect(fileLoader()).toMatchSnapshot();
  });
});

describe('cssLoader', () => {
  test('returns loader config for CSS files', () => {
    expect(cssLoader()).toMatchSnapshot();
  });
  test('returns loader config for CSS files with postcss disabled', () => {
    expect(cssLoader({ postcss: false })).toMatchSnapshot();
  });
  test('returns loader config for CSS files with with inline disabled', () => {
    expect(cssLoader({ hot: false })).toMatchSnapshot();
  });
  test('returns loader config for CSS files with with inline and hot reloading disabled', () => {
    expect(cssLoader({ hot: false, inline: false })).toMatchSnapshot();
  });
});

describe('jsxLoader', () => {
  test('returns loader config for JavaScript files', () => {
    expect(jsxLoader().fragment.module.rules).toEqual([
      {
        exclude: undefined,
        include: undefined,
        test: /\.jsx?$/i,
        use: [
          {
            loader: require.resolve('babel-loader'),
            options: {
              babelrc: true,
              cacheDirectory: true,
              plugins: [],
              presets: [
                [
                  'amex',
                  {
                    modern: true,
                    'preset-env': {
                      modules: false,
                    },
                  },
                ],
              ],
            },
          },
        ],
      },
    ]);
  });
  test('returns loader config for JavaScript files with hot reloading enabled', () => {
    expect(jsxLoader({ hot: true }).fragment.module.rules).toEqual([
      {
        exclude: undefined,
        include: undefined,
        test: /\.jsx?$/i,
        use: [
          {
            loader: require.resolve('babel-loader'),
            options: {
              babelrc: true,
              cacheDirectory: true,
              plugins: [require.resolve('react-refresh/babel')],
              presets: [
                [
                  'amex',
                  {
                    modern: true,
                    'preset-env': {
                      modules: false,
                    },
                  },
                ],
              ],
            },
          },
        ],
      },
    ]);
  });
});

describe('createJavaScriptSourceLoadersConfigFragment', () => {
  test('returns config for javascript script loader', () => {
    expect(createJavaScriptSourceLoadersConfigFragment()).toEqual({
      module: {
        rules: [
          {
            test: /\.(woff|woff2|ttf|eot|svg|png|jpg|jpeg|gif|webm)(\?.*)?$/,
            use: [
              {
                loader: 'file-loader',
                options: {
                  name: 'assets/[name].[ext]',
                },
              },
            ],
          },
          {
            exclude: /node_modules/,
            include: undefined,
            test: /\.(sa|sc|c)ss$/,
            use: [
              {
                loader: 'style-loader',
              },
              {
                loader: 'css-loader',
                options: {
                  importLoaders: 2,
                  modules: {
                    localIdentName: '[name]__[local]___[contenthash:base64:5]',
                  },
                },
              },
              {
                loader: 'sass-loader',
              },
              {
                loader: 'postcss-loader',
                options: {
                  postcssOptions: {
                    plugins: {
                      cssnano: {},
                      'postcss-browser-reporter': {},
                      'postcss-preset-env': {
                        browsers: 'last 2 versions',
                      },
                    },
                    syntax: 'postcss-scss',
                  },
                  sourceMap: undefined,
                },
              },
            ],
          },
          {
            exclude: undefined,
            include: undefined,
            test: /\.jsx?$/i,
            use: [
              {
                loader: require.resolve('babel-loader'),
                options: {
                  babelrc: true,
                  cacheDirectory: true,
                  plugins: [],
                  presets: [
                    [
                      'amex',
                      {
                        modern: true,
                        'preset-env': {
                          modules: false,
                        },
                      },
                    ],
                  ],
                },
              },
            ],
          },
        ],
      },
    });
  });
});
describe('createEsBuildConfigFragment', () => {
  test('returns config for es build', () => {
    expect(createEsBuildConfigFragment()).toEqual({
      module: {
        rules: [
          {
            loader: require.resolve('esbuild-loader'),
            options: {
              loader: 'jsx',
              target: 'es2015',
            },
            test: /\.jsx?$/i,
          },
        ],
      },
    });
  });
});
