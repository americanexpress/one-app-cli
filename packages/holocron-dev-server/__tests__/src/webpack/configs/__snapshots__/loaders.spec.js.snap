// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cssLoader returns loader config for CSS files 1`] = `
Object {
  "fragment": Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "include": undefined,
          "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
          "use": Array [
            Object {
              "loader": "style-loader",
            },
            Object {
              "loader": "css-loader",
              "options": Object {
                "importLoaders": 2,
                "modules": Object {
                  "localIdentName": "[name]__[local]___[contenthash:base64:5]",
                },
              },
            },
            Object {
              "loader": "sass-loader",
            },
            Object {
              "loader": "postcss-loader",
              "options": Object {
                "postcssOptions": Object {
                  "plugins": Object {
                    "cssnano": Object {},
                    "postcss-browser-reporter": Object {},
                    "postcss-preset-env": Object {
                      "browsers": "last 2 versions",
                    },
                  },
                  "syntax": "postcss-scss",
                },
                "sourceMap": undefined,
              },
            },
          ],
        },
      ],
    },
  },
  "rule": Object {
    "exclude": /node_modules/,
    "include": undefined,
    "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
    "use": Array [
      Object {
        "loader": "style-loader",
      },
      Object {
        "loader": "css-loader",
        "options": Object {
          "importLoaders": 2,
          "modules": Object {
            "localIdentName": "[name]__[local]___[contenthash:base64:5]",
          },
        },
      },
      Object {
        "loader": "sass-loader",
      },
      Object {
        "loader": "postcss-loader",
        "options": Object {
          "postcssOptions": Object {
            "plugins": Object {
              "cssnano": Object {},
              "postcss-browser-reporter": Object {},
              "postcss-preset-env": Object {
                "browsers": "last 2 versions",
              },
            },
            "syntax": "postcss-scss",
          },
          "sourceMap": undefined,
        },
      },
    ],
  },
}
`;

exports[`cssLoader returns loader config for CSS files with postcss disabled 1`] = `
Object {
  "fragment": Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "include": undefined,
          "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
          "use": Array [
            Object {
              "loader": "style-loader",
            },
            Object {
              "loader": "css-loader",
              "options": Object {
                "importLoaders": 2,
                "modules": Object {
                  "localIdentName": "[name]__[local]___[contenthash:base64:5]",
                },
              },
            },
            Object {
              "loader": "sass-loader",
            },
            Object {
              "loader": "postcss-loader",
              "options": Object {
                "postcssOptions": Object {
                  "plugins": Object {
                    "cssnano": Object {},
                    "postcss-browser-reporter": Object {},
                    "postcss-preset-env": Object {
                      "browsers": "last 2 versions",
                    },
                  },
                  "syntax": "postcss-scss",
                },
                "sourceMap": undefined,
              },
            },
          ],
        },
      ],
    },
  },
  "rule": Object {
    "exclude": /node_modules/,
    "include": undefined,
    "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
    "use": Array [
      Object {
        "loader": "style-loader",
      },
      Object {
        "loader": "css-loader",
        "options": Object {
          "importLoaders": 2,
          "modules": Object {
            "localIdentName": "[name]__[local]___[contenthash:base64:5]",
          },
        },
      },
      Object {
        "loader": "sass-loader",
      },
      Object {
        "loader": "postcss-loader",
        "options": Object {
          "postcssOptions": Object {
            "plugins": Object {
              "cssnano": Object {},
              "postcss-browser-reporter": Object {},
              "postcss-preset-env": Object {
                "browsers": "last 2 versions",
              },
            },
            "syntax": "postcss-scss",
          },
          "sourceMap": undefined,
        },
      },
    ],
  },
}
`;

exports[`cssLoader returns loader config for CSS files with with inline and hot reloading disabled 1`] = `
Object {
  "fragment": Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "include": undefined,
          "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
          "use": Array [
            Object {
              "loader": "css-loader",
              "options": Object {
                "importLoaders": 2,
                "modules": Object {
                  "localIdentName": "[name]__[local]___[contenthash:base64:5]",
                },
              },
            },
            Object {
              "loader": "sass-loader",
            },
            Object {
              "loader": "postcss-loader",
              "options": Object {
                "postcssOptions": Object {
                  "plugins": Object {
                    "cssnano": Object {},
                    "postcss-browser-reporter": Object {},
                    "postcss-preset-env": Object {
                      "browsers": "last 2 versions",
                    },
                  },
                  "syntax": "postcss-scss",
                },
                "sourceMap": undefined,
              },
            },
          ],
        },
      ],
    },
  },
  "rule": Object {
    "exclude": /node_modules/,
    "include": undefined,
    "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
    "use": Array [
      Object {
        "loader": "css-loader",
        "options": Object {
          "importLoaders": 2,
          "modules": Object {
            "localIdentName": "[name]__[local]___[contenthash:base64:5]",
          },
        },
      },
      Object {
        "loader": "sass-loader",
      },
      Object {
        "loader": "postcss-loader",
        "options": Object {
          "postcssOptions": Object {
            "plugins": Object {
              "cssnano": Object {},
              "postcss-browser-reporter": Object {},
              "postcss-preset-env": Object {
                "browsers": "last 2 versions",
              },
            },
            "syntax": "postcss-scss",
          },
          "sourceMap": undefined,
        },
      },
    ],
  },
}
`;

exports[`cssLoader returns loader config for CSS files with with inline disabled 1`] = `
Object {
  "fragment": Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "include": undefined,
          "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
          "use": Array [
            Object {
              "loader": "style-loader",
            },
            Object {
              "loader": "css-loader",
              "options": Object {
                "importLoaders": 2,
                "modules": Object {
                  "localIdentName": "[name]__[local]___[contenthash:base64:5]",
                },
              },
            },
            Object {
              "loader": "sass-loader",
            },
            Object {
              "loader": "postcss-loader",
              "options": Object {
                "postcssOptions": Object {
                  "plugins": Object {
                    "cssnano": Object {},
                    "postcss-browser-reporter": Object {},
                    "postcss-preset-env": Object {
                      "browsers": "last 2 versions",
                    },
                  },
                  "syntax": "postcss-scss",
                },
                "sourceMap": undefined,
              },
            },
          ],
        },
      ],
    },
  },
  "rule": Object {
    "exclude": /node_modules/,
    "include": undefined,
    "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
    "use": Array [
      Object {
        "loader": "style-loader",
      },
      Object {
        "loader": "css-loader",
        "options": Object {
          "importLoaders": 2,
          "modules": Object {
            "localIdentName": "[name]__[local]___[contenthash:base64:5]",
          },
        },
      },
      Object {
        "loader": "sass-loader",
      },
      Object {
        "loader": "postcss-loader",
        "options": Object {
          "postcssOptions": Object {
            "plugins": Object {
              "cssnano": Object {},
              "postcss-browser-reporter": Object {},
              "postcss-preset-env": Object {
                "browsers": "last 2 versions",
              },
            },
            "syntax": "postcss-scss",
          },
          "sourceMap": undefined,
        },
      },
    ],
  },
}
`;

exports[`fileLoader returns loader config for file types for webpack version 4 and below 1`] = `
Object {
  "fragment": Object {
    "module": Object {
      "rules": Array [
        Object {
          "test": /\\\\\\.\\(woff\\|woff2\\|ttf\\|eot\\|svg\\|png\\|jpg\\|jpeg\\|gif\\|webm\\)\\(\\\\\\?\\.\\*\\)\\?\\$/,
          "use": Array [
            Object {
              "loader": "file-loader",
              "options": Object {
                "name": "assets/[name].[ext]",
              },
            },
          ],
        },
      ],
    },
  },
  "rule": Object {
    "test": /\\\\\\.\\(woff\\|woff2\\|ttf\\|eot\\|svg\\|png\\|jpg\\|jpeg\\|gif\\|webm\\)\\(\\\\\\?\\.\\*\\)\\?\\$/,
    "use": Array [
      Object {
        "loader": "file-loader",
        "options": Object {
          "name": "assets/[name].[ext]",
        },
      },
    ],
  },
}
`;

exports[`fileLoader returns loader config for file types for webpack version 5 and above 1`] = `
Object {
  "fragment": Object {
    "module": Object {
      "rules": Array [
        Object {
          "test": /\\\\\\.\\(woff\\|woff2\\|ttf\\|eot\\|svg\\|png\\|jpg\\|jpeg\\|gif\\|webm\\)\\(\\\\\\?\\.\\*\\)\\?\\$/,
          "type": "asset/resource",
        },
      ],
    },
    "output": Object {
      "assetModuleFilename": "assets/[name].[ext]",
    },
  },
  "rule": Object {
    "test": /\\\\\\.\\(woff\\|woff2\\|ttf\\|eot\\|svg\\|png\\|jpg\\|jpeg\\|gif\\|webm\\)\\(\\\\\\?\\.\\*\\)\\?\\$/,
    "type": "asset/resource",
  },
}
`;
